File: RouterLib/src/main/java/com/kronos/router/interceptor/RealCall.java
package com.kronos.router.interceptor;

import android.content.Context;
import android.os.Bundle;

import com.kronos.router.exception.RouteNotFoundException;
import com.kronos.router.model.HostParams;
import com.kronos.router.model.RouterParams;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class RealCall {

    private final Map<String, HostParams> hostMap;

    public RealCall(Map<String, HostParams> hostMap) {
        this.hostMap = hostMap;
    }

    public void open(String url, Context context, Bundle bundle) throws RouteNotFoundException {
        getParamsWithInterceptorChain(url, context, bundle);
    }

    private void getParamsWithInterceptorChain(String url, Context context, Bundle bundle) throws RouteNotFoundException {
        List<Interceptor> interceptors = new ArrayList<>();
        interceptors.add(new TestInterceptor());
        interceptors.add(new CacheInterceptor());
        interceptors.add(new RouterInterceptor(interceptors));
        Interceptor.Chain chain = new RealInterceptorChain(interceptors, url, hostMap, 0, context, bundle);
        chain.proceed();
    }

}

File: RouterLib/src/main/java/com/kronos/router/interceptor/RealCall.kt
package com.kronos.router.interceptor

import android.content.Context
import com.kronos.router.KRequest
import com.kronos.router.exception.RouteNotFoundException
import com.kronos.router.model.HostParams
import com.kronos.router.model.RouterParams
import com.kronos.router.utils.RouteConfiguration
import java.util.*

class RealCall(private val hostMap: Map<String, HostParams>, private val config: RouteConfiguration) {

    private val cachedRoutes: HashMap<String, RouterParams> = hashMapOf()

    @Throws(RouteNotFoundException::class)
    fun open(request: KRequest, context: Context) {
        getParamsWithInterceptorChain(request, context)
    }

    @Throws(RouteNotFoundException::class)
    private fun getParamsWithInterceptorChain(request: KRequest, context: Context) {
        val interceptors: MutableList<Interceptor> = ArrayList()
        if (config.lazyInitializer) {
            interceptors.add(LazyInitInterceptor())
        }
        interceptors.addAll(config.interceptors)
        interceptors.add(CacheInterceptor(cachedRoutes))
        interceptors.add(RouterInterceptor(cachedRoutes))
        val chain: Interceptor.Chain = RealInterceptorChain(interceptors, request, hostMap, 0, context)
        chain.proceed()
    }

}